feat: Enable aperture modifications without a full mechanics solver

LilianeRA opened the pull request:
This PR is the first step for implementing a simplified hydromechanical coupling based on the Barton–Bandis aperture model and a prescribed stress path.

The main change introduced here is the new permeability constitutive law (BartonBandisPermeability) that internally computes fracture aperture, according to the current pressure and normal stress, before updating its permeability.

------------------------------------------------------------------------------------------------------------------------------
EDIT: as advised, we are reformulating the code design to better encapsulate what is new: the prescribed stress path implementation. Please, don't review the current code until further notice.
------------------------------------------------------------------------------------------------------------------------------

LilianeRA committed:
Submitting the latest version of the code that gives the expected result. A problem sample is defined in 'stressPathDrivenGeomechanics/'. The pressure and aperture data will be saved in the files 'fracPressure_history.hdf5' and 'fracAperture_history.hdf5'.

LilianeRA committed:
Submitting the latest version of the code that gives the expected result. A problem sample is defined in 'stressPathDrivenGeomechanics/'. The pressure and aperture data will be saved in the files 'fracPressure_history.hdf5' and 'fracAperture_history.hdf5'.

LilianeRA committed:
Removing unnecessary files for this branch

LilianeRA committed:
Adding unit test for the oedometric conditions. It still does not work as expected.

LilianeRA committed:
Implementing the Barton-Bandis law version that computes the oedometric trajectory: BartonBandisStressPathDriven.

LilianeRA committed:
Setting the stress computation as a private function; Exclusive UpdatePassThru function for our new constitutive law; Deleting commented function in FlowSolverBase; Using previous implementation of the Barton Bandis law, with Kni

LilianeRA committed:
Setting the stress computation as a private function; Exclusive UpdatePassThru function for our new constitutive law; Deleting commented function in FlowSolverBase; Using previous implementation of the Barton Bandis law, with Kni

LilianeRA committed:
Fracture aperture update implemented in the FlowSolverBase; Not compiling the new solver anymore;

LilianeRA committed:
Fracture aperture update implemented in the FlowSolverBase; Not compiling the new solver anymore;

LilianeRA committed:
Creating input files

LilianeRA committed:
Last commit before the pull request

LilianeRA committed:
Adding simulation tests for comparisson with the analytical solution

LilianeRA committed:
Adding sample file for testing and first version of the RST documenting files,  deleting commented code lines.

LilianeRA committed:
Changing the name of the two optional flags from the FlowSolverBase. Updating the RST documentation.

LilianeRA committed:
Deleting unnecessary directory for compiling latex equation format in the RST documentation; Removing flag for updating stencils; Minor documentation update.

LilianeRA committed:
Testing the constitutive law that relates permeability to pressure. The new fracture aperture is calculated internally, and it is not propagated to the aperture field. For now, the test requires that the pressure difference to be very small.

LilianeRA committed:
Testing the constitutive law that relates permeability to pressure. The new fracture aperture is calculated internally, and it is not propagated to the aperture field. For now, the test requires that the pressure difference to be very small.

LilianeRA committed:
Constitutive contact law for stress-path driven Barton-Bandis deleted. Replacing it by a constitutive relation for permeability. Updated aperture computed internally. Results still validated by analytical solution. Sample file updated.

LilianeRA committed:
Constitutive contact law for stress-path driven Barton-Bandis deleted. Replacing it by a constitutive relation for permeability. Updated aperture computed internally. Results still validated by analytical solution. Sample file updated.

LilianeRA committed:
Reverting unused modifications

jafranc commented on src/CMakeLists.txt:
```diff
@@ -23,6 +23,12 @@ set( CMAKE_ENABLE_EXPORTS ON )
 # populated by `geos_add_test` (private list of all geos tests, keep in sync with the macro definition)
 set_property( GLOBAL PROPERTY geos_tests_exe_list )
 
+set(ENABLE_PYGEOSX ON CACHE BOOL "")
+set(Python3_EXECUTABLE "/home/lilianera/miniconda3/envs/PYGEOS312/bin/python")
+set(Python3_INCLUDE_DIRS "/home/lilianera/miniconda3/envs/PYGEOS312/include/python3.12")
+set(Python3_LIBRARIES "/home/lilianera/miniconda3/envs/PYGEOS312/lib")
+
```
```suggestion
```

LilianeRA replied:
Lines deleted

jafranc commented on .gitignore:
```diff
@@ -20,6 +20,7 @@
 uberenv_libs
 spack-*.txt
 /src/docs/sphinx/datastructure
+/src/docs/sphinx/build
```
```suggestion
```

LilianeRA replied:
Line deleted

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:
```diff
@@ -47,6 +47,7 @@
 #include "physicsSolvers/fluidFlow/kernels/singlePhase/FluidUpdateKernel.hpp"
 #include "physicsSolvers/fluidFlow/kernels/singlePhase/SolidInternalEnergyUpdateKernel.hpp"
 
+//#include "constitutive/contact/HydraulicApertureRelationSelector.hpp"
```
```suggestion
```

LilianeRA replied:
Comment deleted

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.hpp:
```diff
@@ -82,9 +82,13 @@ class FlowSolverBase : public PhysicsSolverBase
 
     static constexpr char const * fluidNamesString() { return "fluidNames"; }
     static constexpr char const * solidNamesString() { return "solidNames"; }
+    //static constexpr char const * stressPathNamesString() { return "stressPathNames"; }
```
```suggestion
```

LilianeRA replied:
Comment deleted

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.cpp:
```diff
@@ -86,20 +87,26 @@ void updatePorosityAndPermeabilityFixedStress( POROUSWRAPPER_TYPE porousWrapper,
 {
   forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOS_DEVICE ( localIndex const k )
   {
+    array1d < real64 > normal(3);
+    normal[0] = normalVector[k][0];
+    normal[1] = normalVector[k][1];
+    normal[2] = normalVector[k][2];
```
Here slicing might be a workable solution

LilianeRA replied:
Ok, I'll push a new version working with arraySlice1d

jafranc commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.cpp:
```diff
@@ -86,20 +87,26 @@ void updatePorosityAndPermeabilityFixedStress( POROUSWRAPPER_TYPE porousWrapper,
+                                                               pressure[k],
+                                                               oldHydraulicAperture[k],
+                                                               newHydraulicAperture[k],
+                                                               normal );
```
```suggestion
                                                               normalVector[k] );
```

LilianeRA replied:
Variable replaced

jafranc commented on src/coreComponents/finiteVolume/SurfaceElementStencil.hpp:
```diff
@@ -333,18 +333,18 @@ SurfaceElementStencilWrapper::
       localIndex const esr1 =  m_elementSubRegionIndices[iconn][k[1]];
       localIndex const ei1  =  m_elementIndices[iconn][k[1]];
-
+      //                                         m_permeability
```
```suggestion

```

LilianeRA replied:
Comment removed

jafranc commented on src/coreComponents/constitutive/permeability/PermeabilityBase.hpp:
```diff
@@ -57,13 +57,15 @@ class PermeabilityBaseUpdate
   }
 
   GEOS_HOST_DEVICE
```
for consistency maybe we want to add a new one but keep the previous

LilianeRA replied:
Ok, I'll add again the code of "updateFromAperture" function, even though it will never be executed. Since we are planning to create a new module for computing analytical mechanics, this class' code will revert to its original version.

jafranc commented on examples/functions/function_examples/function_examples.xml:
```diff
@@ -5,15 +5,15 @@
     symbolic math function should expect the variables to be given, and an expression that it will compile.
     A list of available functions is given here: https://github.com/kobalicek/mathpresso -->
-    <SymbolicFunction
+    <!--<SymbolicFunction
```
? why is this now commented out ?

LilianeRA replied:
Whenever I compile GEOS, even in the main branch, I get the error:

GEOS/examples/functions/function_examples/function_examples.xml:11: Schemas validity error : Element 'SymbolicFunction': This element is not expected. Expected is one of ( MultivariableTableFunction, TableFunction ).
make[2]: *** [CMakeFiles/geosx_validate_all_xml_files.dir/build.make:71: CMakeFiles/geosx_validate_all_xml_files] Error 1
make[1]: *** [CMakeFiles/Makefile2:2537: CMakeFiles/geosx_validate_all_xml_files.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

It made me believe that the compilation failed, but the binary file was perfectly generated. I commented it so I could see better if my modifications crashed or not. This line can be executed again, no problem.

jafranc commented on host-configs/ubuntu.cmake:
```diff
@@ -0,0 +1,32 @@
+set( CONFIG_NAME "ubuntu" ) 
```
I think `environment.cmake` is doing pretty much the same

LilianeRA replied:
I copied the file 'quick-start-template.cmake' and edited it to work in WSL. Not required for the PR proposal, its just a configuration file for my PC. It can be deleted for now

jafranc commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.cpp:
```diff
@@ -0,0 +1,102 @@
+{
+  PermeabilityBase::postInputInitialization();
+
+  if( m_transversalPermeability > -1 )
```
shouldn't it be weakly positive ?

LilianeRA replied:
m_transversalPermeability is initialized as -1. So I believe -1 is a flag. Currently there is no XML sample file that sets this parameter, therefore this condition is never true.

jafranc commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,278 @@
+
+    // Computation of maximum fracture closure (Barton-Bandis parameter)
+    // Fracture traction via Terzaghi's Principle
+    real64 const sigma_c0[3] = { m_referenceTotalStress[0] * normal[0] - biot_pressure * normal[0],
```
why not leveraging `LvArray::tensorOp` ?

LilianeRA replied:
You're right, its better to use tensorOps. I'll switch it.

jafranc commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,278 @@
+    real64 const sigma_n0 = sigma_c0[0]*normal[0] + 
+                            sigma_c0[1]*normal[1] + 
+                            sigma_c0[2]*normal[2];
+    real64 const g0 = (-m_normalStiffness*m_aperture0 + 
+                        std::sqrt((m_normalStiffness*m_aperture0)*
+                        (m_normalStiffness*m_aperture0) + 
+                        4.0*m_normalStiffness*sigma_n0*m_aperture0)) / (2.0*m_normalStiffness);
```
isn't it (Kni*a0)*(sqrt(Kni*a0)-1 + 4*sigma_n0) / (2*Kni) ? factorized that would be easier to read

LilianeRA replied:
I agree, its hard to read. The original equation is \dfrac{-K_{ni}a_0 + \sqrt{(K_{ni}a_0)^2 + 4K_{ni}a_0\sigma_{n0}}}{2K_{ni}}. I'll factorized it.

jafranc commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,278 @@
+    real64 const newHydraulicAperture = maximumFractureClosure - fractureClosure;
+
+    // derivative
+    real64 const Kni_apert = m_normalStiffness*maximumFractureClosure;
+    real64 const Kni_aper_stress = Kni_apert + normalComponentOfStressOnFracture;
+    dAperture_dStress = -(Kni_apert*maximumFractureClosure) / (Kni_aper_stress * Kni_aper_stress);
+
```
please review variable's name

LilianeRA replied:
I believe you mean the variables must be in 'camelCase'. I'll correct them.

jafranc commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,278 @@
+  GEOS_FORCE_INLINE
+  real64 computeFractureStress( real64 const pressure, array1d< real64 > const & normal, real64 & dStress_dPressure ) const
+  {  
+    // TODO: remove this lambda expression
+    auto matmul = [](real64 const (&u)[3], array1d< real64 > const &v, array1d< real64 > &r) -> void
+    {
+      r[0] = u[0]*v[0];
+      r[1] = u[1]*v[1];
+      r[2] = u[2]*v[2];
+    }; 
```
```suggestion
```

LilianeRA replied:
After reading geometricTensorOps.hpp, I'll replace this with LvArray::hadamardProduct

LilianeRA committed:
Modifications suggested by Jacques Franc

LilianeRA committed:
Merge branch 'develop' into feature/prescribedStressPath_constitutive

dkachuma commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,323 @@
+    /// string/key for aperture under zero normal stress
+    static constexpr char const * apertureZeroString() { return "referenceAperture"; }
+    static constexpr char const * biotString()                  { return "biot"; }
+    static constexpr char const * poissonString()               { return "poisson"; }
```
```suggestion
    static constexpr char const * biotString()                  { return "biotCoefficient"; }
    static constexpr char const * poissonString()               { return "poissonRatio"; }
```

dkachuma commented on src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp:
```diff
@@ -0,0 +1,323 @@
+
+private:
+
+  int m_numDimensionsToUpdate;
```
```suggestion
  int const m_numDimensionsToUpdate;
```

LilianeRA replied:
I could define this variable as constant, but since we are already planning another code design approach that does not involve this permeability class, I will leave it as it is for now.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.hpp:
```diff
@@ -82,9 +82,13 @@ class FlowSolverBase : public PhysicsSolverBase
     static constexpr char const * thermalConductivityNamesString() { return "thermalConductivityNames"; }
+
+    static constexpr char const * computePrescribedStressPathString() { return "computePrescribedStressPath"; }
+    static constexpr char const * hydraulicApertureRelationNameString() { return "hydraulicApertureRelationName"; }
```
```suggestion
```
I suppose the use of these has been removed. Perhaps we don't need them anymore.

dkachuma commented on src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.cpp:
```diff
@@ -85,21 +86,32 @@ void updatePorosityAndPermeabilityFixedStress( POROUSWRAPPER_TYPE porousWrapper,
   } );
 }
 
+void nada(arraySlice1d< real64 const > const & normal)
```
I'm guessing this is debug code that needs to be removed?

LilianeRA replied:
Yes, you're right. Code removed.

dkachuma commented on src/coreComponents/constitutive/solid/CompressibleSolid.cpp:
```diff
@@ -48,13 +49,15 @@ typedef CompressibleSolid< PressurePorosity, CarmanKozenyPermeability > Compress
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, CompressibleRockCK, string const &, Group * const )
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, CompressibleRockPressurePerm, string const &, Group * const )
 REGISTER_CATALOG_ENTRY( ConstitutiveBase, FractureRock, string const &, Group * const )
+REGISTER_CATALOG_ENTRY( ConstitutiveBase, FractureRockBB, string const &, Group * const )
```
I can't find `FractureRockBB`

LilianeRA replied:
Sorry, this code is a remnant from the previous implementation, when our method was implemented as a constitutive law of contact. I'll delete it.

dkachuma commented on src/coreComponents/constitutive/permeability/ParallelPlatesPermeability.hpp:
```diff
@@ -60,26 +60,30 @@ class ParallelPlatesPermeabilityUpdate : public PermeabilityBaseUpdate
     compute( oldHydraulicAperture,
              newHydraulicAperture,
              dHydraulicAperture_dNormalJump,
              m_permeability[k][0],
              m_dPerm_dDispJump[k][0] );
```
```suggestion
m_permeability[k][q],
m_dPerm_dDispJump[k][q] );
```
I would still use the quadrature index although it's always 0. Just for posterity.

dkachuma reviewed (commented):
Can you please remove any debug/dead code and sync to `develop`. Things have moved quite a bit making this a little harder to review.

LilianeRA committed:
Update src/coreComponents/physicsSolvers/fluidFlow/FlowSolverBase.hpp

Removing unused optional flags for input file. Remnants of when the proposed method was modeled as a contact constituive class.

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

LilianeRA committed:
Update src/coreComponents/constitutive/permeability/BartonBandisPermeability.hpp

Better variable name description

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

LilianeRA committed:
Update src/coreComponents/constitutive/permeability/ParallelPlatesPermeability.hpp

Suggestion by Kachuma: "I would still use the quadrature index although it's always 0. Just for posterity."

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

LilianeRA committed:
Updating variable names of biot coefficient and poisson ratio; Deleting dead code;

LilianeRA committed:
Updating variable names of biot coefficient and poisson ratio; Deleting dead code;

LilianeRA committed:
Removing debug/dead code

LilianeRA committed:
Fixing errors during compilation, and merge mistakes after rebase.

LilianeRA committed:
Updating the base code and adding BartonBandisPemeability class. Results change slightly in the first few iterations.

LilianeRA committed:
Adding documentation

LilianeRA committed:
Merge branch 'develop' into feature/prescribedStressPath_constitutive
